pkg/users/local.js

```diff
@@ -47,60 +48,76 @@ const sortGroups = groups => {
     const shadow = useFile("/etc/shadow", { superuser: true });
     const groups = useFile("/etc/group", { syntax: etc_group_syntax });
     const current_user_info = useLoggedInUser();
 
     const [details, setDetails] = useState(null);
-    useEffect(() => {
-        getLogins().then(setDetails);
+    useInit(() => {
+        getLogins(shadow).then(setDetails);
 
         // Watch `/var/run/utmp` to register when user logs in or out
         const handle = cockpit.file("/var/run/utmp", { superuser: "try", binary: true });
         handle.watch(() => {
-            getLogins().then(setDetails);
+            getLogins(shadow).then(setDetails);
         });
-        return handle.close;
-    }, [path, accounts, shadow]);
+        return handle;
+    }, [shadow], null, handle => handle.close());
 
     // lastlog uses same sorting as /etc/passwd therefore arrays can be combined based on index
-    let accountsInfo = [];
-    if (accounts && details)
-        accountsInfo = accounts.map((account, i) => {
-            return Object.assign({}, account, details[i]);
-        });
+    const accountsInfo = useMemo(() => {
+        if (accounts && details)
+            return accounts.map((account, i) => {
+                return Object.assign({}, account, details[i]);
+            });
+        else
+            return [];
+    }, [accounts, details]);
 
-    const groupsExtraInfo = sortGroups(
+    const groupsExtraInfo = useMemo(() => sortGroups(
         (groups || []).map(group => {
             const userlistPrimary = accountsInfo.filter(account => account.gid === group.gid).map(account => account.name);
             const userlist = group.userlist.filter(el => el !== "");
             return ({ ...group, userlistPrimary, userlist, members: userlist.length + userlistPrimary.length, isAdmin: admins.includes(group.name) });
         })
-    );
+    ), [groups, accountsInfo]);
 
     if (path.length === 0) {
-        return <AccountsMain accountsInfo={accountsInfo} current_user={current_user_info && current_user_info.name} groups={groupsExtraInfo || []} />;
-    } else {
         return (
-            <AccountDetails accounts={accountsInfo} groups={groupsExtraInfo || []} shadow={shadow || []}
+            <AccountsMain
+                accountsInfo={accountsInfo}
+                current_user={current_user_info && current_user_info.name}
+                groups={groupsExtraInfo}
+                isGroupsExpanded={isGroupsExpanded}
+                setIsGroupsExpanded={setIsGroupsExpanded}
+            />
+        );
+    } else if (path.length === 1) {
+        return (
+            <AccountDetails accounts={accountsInfo} groups={groupsExtraInfo} shadow={shadow || []}
                             current_user={current_user_info && current_user_info.name} user={path[0]} />
         );
-    }
+    } else if (path.length === 2 && path[0] === "group" && groupsExtraInfo.length > 0 && accountsInfo.length > 0) {
```

Comment:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18136-20230125-130546-9f4aa1e5-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)